icefoganalytics/internal-data-portal
# Implement Organizational Heirarchy Data

Fixes https://github.com/icefoganalytics/internal-data-portal/issues/18

Relates to:
- https://github.com/icefoganalytics/internal-data-portal/issues/7

# Context

**Is your feature request related to a problem? Please describe.**
The datasets new page currently uses fake data for department/division/branch/unit (organizational hierarchy data), this needs to be fixed.

Probably add the user sync as well, since I only need to copy/paste that from the ["travel auth" project](https://github.com/icefoganalytics/travel-authorization).

**Describe the solution you'd like**
Add a `orgnizational_hierarchies` table with:
- department
- division
- branch
- unit

and tie it to the appropriate tables.

**Additional context**
Data should track last synced time, provide some interface to sync, and/or be synced on app boot.
I might defer auto-sync for later, and just implement a sync button and sync on boot, at most once a day, for now.

# Implementation

Add ability to sync user with Yukon Government directory.
Add ability to sync user groups with Yukon Government directory.

# Screenshots

http://localhost:8080/dashboard
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/cf20d1f1-4f7a-438b-a98d-2a3650d2a90c)

http://localhost:8080/datasets/new - Empty
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/4b37e334-a2cd-4224-bd6e-3a99ef594fb6)

http://localhost:8080/datasets/new - Filled in
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/fd9cf815-25ce-40fa-9574-f5713c208131)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. Once you are redirected to the http://localhost:8080/dashboard page, click on the "Sync User Groups" button to load the user groups from the Yukon government directory. It takes about 8 seconds.
5. Next click the "Create Dataset" button.
6. Look at the fields, and you will see that many of them are disabled.
7. Fill in the owner name to unlock and prefill the owner position, department, division, branch, and unit info.
8. Fill in the support name to unlock and prefill the support email, and support position fields.
9. Check that when you select a different department, division, or branch, that it clears all the lower tier groups.
10. Check that you can save.
